R/utils.R

```diff
@@ -326,3 +327,86 @@ h_confint <- function(x, parm, level = 0.95, transform, include_se = FALSE, ...)
   }
   transform(ret)
 }
+
+#' Check Whether Two Factors are Nested
+#'
+#' This function checks whether two factor variables are nested in terms of their levels
+#' and the mapping of observations to these levels:
+#'
+#' - Both factors must have the same length.
+#' - The positions of `NA` values must be identical in both factors.
+#' - After removing `NA` values, `f1` must have at least as many observed levels as `f2`.
+#' - The mapping of observations to levels must be consistent, meaning that each level in `f1`
+#'   corresponds to exactly one level in `f2`. On the other hand, multiple levels in `f1`
+#'   can map to the same level in `f2`.
+#'
+#' @param f1 A factor variable, supposed to be more fine grained than `f2`.
+#' @param f2 A factor variable, supposed to be more coarse grained than `f1`.
+#' @return `TRUE` if `f1` is nested in `f2`, `FALSE` otherwise.
+#'
+#' @keywords internal
+h_are_factors_nested <- function(f1, f2) {
+  assert_factor(f1)
+  assert_factor(f2)
+
+  if (length(f1) != length(f2)) {
+    return(FALSE)
+  }
+
+  f1_na <- which(is.na(f1))
+  f2_na <- which(is.na(f2))
+  if (!identical(f1_na, f2_na)) {
+    return(FALSE)
+  }
+
+  f1 <- droplevels(stats::na.omit(f1))
+  f2 <- droplevels(stats::na.omit(f2))
+  if (nlevels(f1) < nlevels(f2)) {
+    return(FALSE)
+  }
+
+  f1_num <- as.numeric(f1)
+  f2_num <- as.numeric(f2)
+  f1_f2_num <- paste(f1_num, f2_num, sep = "-")
+  unique_combinations <- unique(f1_f2_num)
+  length(unique_combinations) == nlevels(f1)
+}
+
+#' Check Unbiased Means of Residuals Across Randomization Strata
+#'
+#' This function checks whether the means of residuals are approximately zero
+#' across specified randomization strata.
+#'
+#' @param residuals_per_group (`numeric`) A named numeric vector containing residuals for each treatment group.
+#' @param df (`data.frame`) The data frame containing the `treatment` and randomization strata variables, produced by
+#'   [h_derived_outcome_vals()] or [h_strat_derived_outcome_vals()].
+#' @param randomization_strata (`character`) A character vector of names of the randomization strata variables in `df`.
+#' @param eps (`numeric`) A small tolerance value to determine if means are close to zero.
+#'
+#' @return `TRUE` if the means of residuals across randomization strata are within the specified tolerance,
+#'   `FALSE` otherwise.
+#'
+#' @keywords internal
+h_unbiased_means_across_strata <- function(
```

Comment:
hi @danielinteractive I think there is a similar function for this feature . can we harmonize them? that function should be `bias` in bias.R